File: be/src/main/java/com/example/onlinebutik/service/ItemService.java
package com.example.onlinebutik.service;


import com.example.onlinebutik.entity.Items;
import com.example.onlinebutik.entity.Users;
import com.example.onlinebutik.model.CartModel;
import com.example.onlinebutik.model.ItemsModel;
import com.example.onlinebutik.repository.IItemRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

@Service
public class ItemService implements  IItemService {

    private static final Double RANGESIMILAR = 200.00;

    @Autowired
    private IItemRepository itemRepository;
    @Autowired
    private AutoMapperService autoMapperService;

    @Override
    public ItemsModel findById(String id) {
        if(itemRepository.findById(id).isPresent())
            return autoMapperService.map(itemRepository.findById(id).get(), ItemsModel.class);
        return new ItemsModel();
    }

    @Override
    public List<ItemsModel> findAllByCart(CartModel cart) {
        List<ItemsModel> list = new ArrayList<ItemsModel>();
        cart.getItems().forEach(cartItemModel -> {
            if(itemRepository.findById(cartItemModel.getItemId()).isPresent())
                list.add(autoMapperService.map(
                        itemRepository.findById(cartItemModel.getItemId()).get(), ItemsModel.class));
        });
        return list.stream().distinct().collect(Collectors.toList());
    }

    @Override
    public ItemsModel updateItemAmount(String itemId, String size, Integer amount) {
        Items i = autoMapperService.map(this.findById(itemId), Items.class);
        i.getAmount().forEach(s->{
            if (s.getName().equals(size)){
                s.setAmount(s.getAmount()-amount);
            }
        });
        return autoMapperService.map(itemRepository.save(i), ItemsModel.class);
    }

    @Override
    public List<ItemsModel> getSimilarItems(String id) {
        ItemsModel item = this.findById(id);
        Double gt = item.getPrice() - ItemService.RANGESIMILAR;
        Double lt = item.getPrice() + ItemService.RANGESIMILAR;
        List<Items> items = itemRepository.getAllByPriceBetween(item.getCategory(), item.getType(), gt, lt);
        List<ItemsModel> models = new ArrayList<ItemsModel>(); ;
        items.forEach(i -> {
            models.add(autoMapperService.map(i, ItemsModel.class));
        });
        return models;
    }

    @Override
    public List<ItemsModel> getFavouriteItems(Users user) {
        List<ItemsModel> items = new ArrayList<>();
        user.getFavouriteItems().forEach(itemId -> {
            ItemsModel i = this.findById(itemId);
            items.add(i);
        });
        return items;
    }

    @Override
    public ItemsModel updateItemStars(Items item) {
        return autoMapperService.map(itemRepository.save(item), ItemsModel.class);
    }

    @Override
    public ItemsModel findBySlug(String slug) {
        return autoMapperService.map(itemRepository.findBySlug(slug), ItemsModel.class);
    }

    @Override
    public List<ItemsModel> findAllByCategoryAndType(String category, String type) {
        List<Items> items = itemRepository.findAllByCategoryAndType(category, type);
        List<ItemsModel> models = new ArrayList<ItemsModel>(); ;
        items.forEach(item -> {
           models.add(autoMapperService.map(item, ItemsModel.class));
        });
        return models;
    }

    public List<ItemsModel>findAll() {
        List<Items> items = itemRepository.findAll();
        List<ItemsModel> models = new ArrayList<ItemsModel>(); ;
        items.forEach(item -> {
            models.add(autoMapperService.map(item, ItemsModel.class));
        });
        return models;
    }

}

File: fe/src/app/services/item.service.ts
import { HttpClient, HttpParams, HttpResponse } from "@angular/common/http";
import { Injectable } from "@angular/core";
import { Observable } from "rxjs";
import { User } from "./user.service";

export interface Size {
    name: string;
    amount: number;

}


export interface Item {
    id: string;
    name: string;
    image: string;
    category: string;
    type: string;
    amount: Size[];
    stars: number;
    price: number;
    description: string;
    country: string,
    slug: string
}

@Injectable()
export class ItemService {

    constructor(private http: HttpClient) {
        
    }

    public findAllByCategoryAndType(category: string, type: string) : Observable<HttpResponse<Item[]>> { 
        let data = new HttpParams().set("category" , category).set("type", type);
        return this.http.get<Item[]>("http://localhost:8080/items/findallbycategoryandtype", {observe: "response", params: data});
    }

    public findBySlug(slug: string) : Observable<HttpResponse<Item>>{
        let data = new HttpParams().set("slug" , slug);
        return this.http.get<Item>("http://localhost:8080/items/findbyslug/"+slug, {observe: "response"});
    }

    public findById(id: string) : Observable<HttpResponse<Item>>{
        return this.http.get<Item>("http://localhost:8080/items/findbyid/"+id, {observe: "response"});
    }

    public getFavouriteItems(user: User) : Observable<HttpResponse<Item[]>>{
        return this.http.post<Item[]>("http://localhost:8080/items/getfavouriteitems", user, {observe: "response"});
    }

    public updateItemAmount(  itemId: string, size: string,  amount: number): Observable<HttpResponse<Item>> {
        let data = {
            "itemId": itemId,
            "amount": amount,
            "size": size
        }
         return this.http.post<Item>("http://localhost:8080/items/updateitemamount", data , {observe: "response"});
    }

    public updateItemStars(  item: Item ): Observable<HttpResponse<Item>> {
         return this.http.post<Item>("http://localhost:8080/items/updateitemstars", item , {observe: "response"});
    }

    public getSimilarItems(item: Item) : Observable<HttpResponse<Item[]>> {
        return this.http.get<Item[]>("http://localhost:8080/items/getsimilaritems/"+item.id, {observe: "response"});
    }
}